### PR TITLE
feat: allow initializing regions in background

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

1. Introduce `initialize_region_in_background`, allowing users to initialize all regions in the background without blocking the startup.
2. Introduce the `RegionNotReady` status code, and non-create/open requests trying to access the region under creating or opening will return the `RegionNotReady` instead of `RegionNotFound.`
3. Introduce the `RegionBusy` status code, ensuring only one exclusive operation(e.g., open region, close region) will be executed simultaneously.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Review by waynexia — APPROVED** [review]

File: src/datanode/src/datanode.rs [hunk 1]
```diff
@@ -233,8 +234,18 @@ impl DatanodeBuilder {
             .new_region_server(log_store, region_event_listener)
             .await?;
 
-        self.initialize_region_server(&region_server, kv_backend, !controlled_by_metasrv)
-            .await?;
+        let region_server_moved = region_server.clone();
+
+        // Opens regions in background.
+        common_runtime::spawn_bg(async move {
```
Comment (fengjiachun) [inline src/datanode/src/datanode.rs:240]: Is the error be ignored?

**Comment by WenyXu** [comment]
@waynexia The new commits changed a lot; it is better to review them again😇

File: src/datanode/src/region_server.rs [hunk 2]
```diff
@@ -306,48 +348,109 @@ impl RegionServerInner {
...
+                    RegionEngineWithStatus::Registering(_) => {
+                        return error::RegionNotReadySnafu { region_id }.fail()
+                    }
+                    RegionEngineWithStatus::Deregistering(_) => {
+                        return error::RegionNotFoundSnafu { region_id }.fail()
+                    }
+                    RegionEngineWithStatus::Ready(engine) => engine,
+                },
+                None => return error::RegionNotFoundSnafu { region_id }.fail(),
+            },
```
Comment (WenyXu) [inline src/datanode/src/region_server.rs:391]: Maybe I should move `region change` related logic into separate methods.

File: src/cmd/src/standalone.rs [hunk 3]
```diff
@@ -93,6 +93,7 @@ impl SubCommand {
 pub struct StandaloneOptions {
     pub mode: Mode,
     pub enable_telemetry: bool,
+    pub initialize_region_in_background: bool,
```
Comment (waynexia) [inline src/cmd/src/standalone.rs:96]: Is it necessary to export this as an option? Or we can set it to `true` as the default behavior meets most requirements and scenario

File: src/datanode/src/datanode.rs [hunk 4]
```diff
@@ -233,8 +234,23 @@ impl DatanodeBuilder {
...
+            kv_backend,
+            !controlled_by_metasrv,
+            node_id,
+        );
+
+        if self.opts.initialize_region_in_background {
+            // Opens regions in background.
+            common_runtime::spawn_bg(async move {
+                if let Err(err) = open_all_regions.await {
+                    error!(err; "Failed to opening regions during the startup.");
```
Comment (MichaelScofield) [inline src/datanode/src/datanode.rs:248]:
```suggestion
                    error!(err; "Failed to open regions during the startup.");
```

File: src/datanode/src/datanode.rs [hunk 5]
```diff
@@ -506,6 +481,66 @@ impl DatanodeBuilder {
...
+    region_server: RegionServer,
+    kv_backend: KvBackendRef,
+    open_with_writable: bool,
+    node_id: u64,
+) -> Result<()> {
+    let datanode_table_manager = DatanodeTableManager::new(kv_backend.clone());
+    let mut regions = vec![];
+    let mut table_values = datanode_table_manager.tables(node_id);
+
+    while let Some(table_value) = table_values.next().await {
```
Comment (MichaelScofield) [inline src/datanode/src/datanode.rs:495]: Is fetching to-be-opened regions suitable placed here if this function can be executed in background? I mean, if here goes wrong, it's pointless to do any background region open jobs.

File: src/datanode/src/error.rs [hunk 6]
```diff
@@ -210,6 +210,12 @@ pub enum Error {
         location: Location,
     },
 
+    #[snafu(display("RegionId {} not ready", region_id))]
```
Comment (MichaelScofield) [inline src/datanode/src/error.rs:213]:
```suggestion
    #[snafu(display("Region {} not ready", region_id))]
```

File: src/datanode/src/region_server.rs [hunk 7]
```diff
@@ -249,9 +250,50 @@ impl FlightCraft for RegionServer {
     }
 }
 
+#[derive(Clone)]
+enum RegionEngineWithStatus {
+    // A opening, or creating region.
```
Comment (MichaelScofield) [inline src/datanode/src/region_server.rs:255]:
```suggestion
    // An opening, or creating region.
```

File: src/datanode/src/region_server.rs (hunk 2 above)
Comment (MichaelScofield) [inline src/datanode/src/region_server.rs:391]:
```suggestion
            RegionChange::None => match region_status {
                Some(RegionEngineWithStatus::Ready(engine)) => engine.clone()
                _ => return error::RegionNotFoundSnafu { region_id }.fail(),
            },
```

File: src/datanode/src/region_server.rs [hunk 8]
```diff
@@ -306,48 +348,109 @@ impl RegionServerInner {
...
-                info!("Region {region_id} is deregistered from engine {engine_type}");
-                self.region_map
-                    .remove(&region_id)
-                    .map(|(id, engine)| engine.set_writable(id, false));
-                self.event_listener.on_region_deregistered(region_id);
+            Err(err) => {
+                // Removes the RegionStatus if the operation fails.
+                match region_change {
+                    RegionChange::None => {}
+                    RegionChange::Register(_) | RegionChange::Deregisters => {
```
Comment (MichaelScofield) [inline src/datanode/src/region_server.rs:445]: Can it simplify the code if we add an `oneshot` channel to `RegionOpenRequest` to indicate the region open result, then handle the result separately, avoiding this long and complex "match"?

File: src/datanode/src/datanode.rs (hunk 5 above)
Comment (WenyXu) [reply src/datanode/src/datanode.rs:495]: Sorry, I didn't get it. Do you meaning that we should move it out of the background task?

Comment (MichaelScofield) [reply src/datanode/src/datanode.rs:495]: yes

File: src/datanode/src/region_server.rs (hunk 2 above)
Comment (WenyXu) [reply src/datanode/src/region_server.rs:391]: Unfortunately, we can't do that; it needs to let the caller know the region is not ready.

File: src/cmd/src/standalone.rs (hunk 3 above)
Comment (fengjiachun) [reply src/cmd/src/standalone.rs:96]: I believe the Standalone mode doesn't need this option and should not initialize in the background.

File: src/datanode/src/region_server.rs (hunk 8 above)
Comment (WenyXu) [reply src/datanode/src/region_server.rs]: I decided to move them into separate methods.

**Comment by WenyXu** [comment]
@MichaelScofield @fengjiachun PTAL

**Comment by MichaelScofield** [comment]
@WenyXu conflicts

**Comment by WenyXu** [comment]
@MichaelScofield @evenyag PTAL

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**